org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TaskNodeViewerSorter.java

```diff
@@ -78,25 +82,30 @@ public int compare(Viewer viewer, Object leftNode, Object rightNode) {
      * @return the new sorter instance
      */
     public static TaskNodeViewerSorter createFor(TaskViewState state) {
-        Ordering<ProjectNode> projectOrdering = createProjectNodeOrdering();
+        Ordering<ProjectNode> projectOrdering = createProjectNodeOrdering(state.isProjectHierarchyFlattened());
         Ordering<TaskNode> taskOrdering = createTaskNodeOrdering(state.isSortByType(), state.isSortByVisibility());
         return new TaskNodeViewerSorter(projectOrdering, taskOrdering);
     }
 
-    private static Ordering<ProjectNode> createProjectNodeOrdering() {
+    private static Ordering<ProjectNode> createProjectNodeOrdering(final boolean isHierarchyFlattened) {
         return new Ordering<ProjectNode>() {
 
             @Override
             public int compare(ProjectNode left, ProjectNode right) {
-                EclipseProject leftRoot = HierarchicalElementUtils.getRoot(left.getEclipseProject());
-                EclipseProject rightRoot = HierarchicalElementUtils.getRoot(right.getEclipseProject());
-                if (leftRoot == rightRoot) {
-                    // do not change sorting of projects that belong to the same root
-                    return 0;
-                } else {
-                    // projects that do not belong to the same root should be grouped by the name of
-                    // their root projects
-                    return leftRoot.getName().compareTo(rightRoot.getName());
+                if (!isHierarchyFlattened) {
+                    return left.toString().compareTo(right.toString());
```

Comment:
❌Should be `leftRoot.getName().compareTo(rightRoot.getName())`.

Reply:
Hi @donat , I am curious about this point. Since we grouped child nodes into parent node folder. Child project will only compare to the node that belongs to the same parent. Why we need to sort them by their parent (nothing to be sorted)?

Reply:
My mistake. What I meant to write is to sort the nodes by the displayed name. I'd do the following:
- Define `ProjectNode.getDisplayName()` and move the implementation there from `TaskNameLabelProvider.getProjectText()`
- Here, return `left.getDisplayName().compareTo(right.getDisplayName())`